### ISSUE TITLE
RollingWindow on DataFrame

### ISSUE THREAD

**Issue by lgomez** [data-forge/data-forge-ts#8]
Hello again,

If you have a minute, would you mind helping me figure out why I can't get `rollingWindow` to do what I expect?

I have a bunch of data formatted like so:
```table
__index__   BTC      ETH
----------  -------  -------
2017-06-15  2261.25  333.499
2017-06-16  2481.74  371.358
2017-06-17  2591.41  373.975
2017-06-18  2568.92  377.562
2017-06-19  2578.25  371.813
2017-06-20  2657.32  376.359
2017-06-21  2709.33  338.292
2017-06-22  2714.58  340.326
2017-06-23  2760.61  340.811
2017-06-24  2688.44  334.167
2017-06-25  2627.95  321.283
2017-06-26  2476.72  262.385
2017-06-27  null     null
2017-06-28  2539.78  301.789
2017-06-29  2536.94  315.543
2017-06-30  2547.95  305.518
2017-07-01  2507.89  282.877
2017-07-02  2445.84  269.758
2017-07-03  2519.06  288.076
2017-07-04  2618.16  284.342
```
And I'm running the following:
```javascript
const performance = df.rollingWindow(2).select(window => {
  return ((window.last() - window.first()) / window.first())
})
.withIndex(pair => pair[0])
.select(pair => pair[1])
console.log(performance.toString());
```
But I keep getting...
```
__index__  __value__
---------  ---------
0          NaN
1          NaN
2          NaN
3          NaN
4          NaN
5          NaN
6          NaN
7          NaN
8          NaN
9          NaN
10         NaN
// truncated for brevity
353        NaN
354        NaN
355        NaN
356        NaN
357        NaN
358        NaN
359        NaN
360        NaN
361        NaN
362        NaN
363        NaN
```
Not sure what I'm doing wrong. I've tried several variations based on the examples on the Guide.

Thank you!

**Comment by ashleydavis** [comment]
Yes I see the problem (although I haven't tested the code myself).

You are doing this:

```
const performance = df.rollingWindow(2).select(window => {
  return ((window.last() - window.first()) / window.first()) // Return the value.
})
.withIndex(pair => pair[0]) // This doesn't work!
.select(pair => pair[1]) // This doesn't work!
console.log(performance.toString());
```

Instead you probably want this:

```
const performance = df.rollingWindow(2).select(window => {
  return [window.getIndex().last(), ((window.last() - window.first()) / window.first())]; // Note returning a _pair_ containing the index and the value.
})
.withIndex(pair => pair[0]) // Extract index.
.select(pair => pair[1]) // Extract value.
console.log(performance.toString());

The reason you include the pair is if you want to reintegrate the new dataframe with the old one, like this:

```
const performance = originalDf.rollingWindow(2).select(window => {
  return [window.getIndex().last(), ((window.last() - window.first()) / window.first())]; // Note returning a _pair_ containing the index and the value.
})
.withIndex(pair => pair[0]) // Extract index.
.select(pair => pair[1]) // Extract value.
const merged = orignalDf.withSeries({ Performance: performance });
console.log( merged.toString());
```

Merging the new series into the original dataframe get's you the combined results.

If you don't want to merge you can simplify your code to this:

```
const performance = df.rollingWindow(2).select(window => {
  return ((window.last() - window.first()) / window.first()) // Only need the value, not the index.
});
console.log(performance.toString());
```

**Comment by lgomez** [comment]
Thank you. Those seem like what I need. But that is the simplest case amongst the frame calculations I need to run. I think the problem I'm having (or what would help me most) is understanding how to run these on a DataFrame so every series (column) is transformed. My looping through each column does not work.

So, for example, given the following DataFrame subset (truncated for simplicity):
```
__index__   ADA   BCH   BTC      BTG   BTS       DASH     EOS   ETC      ETH
----------  ----  ----  -------  ----  --------  -------  ----  -------  -------
2017-06-16  null  null  2481.74  null  0.318329  165.654  null  18.4334  371.358
2017-06-17  null  null  2591.41  null  0.349656  169.012  null  21.3327  373.975
2017-06-18  null  null  2568.92  null  0.335961  177.434  null  22.1432  377.562
2017-06-19  null  null  2578.25  null  0.330724  198.481  null  22.9697  371.813
2017-06-20  null  null  2657.32  null  0.331358  203.353  null  21.6293  376.359
2017-06-21  null  null  2709.33  null  0.32211   185.504  null  20.4588  338.292
2017-06-22  null  null  2714.58  null  0.312718  183.122  null  20.4527  340.326
2017-06-23  null  null  2760.61  null  0.324769  186.924  null  21.3475  340.811
2017-06-24  null  null  2688.44  null  0.327701  185.133  null  20.2919  334.167
2017-06-25  null  null  2627.95  null  0.312531  181.87   null  22.5758  321.283
```
When I run:
```javascript
const performance = subset.proportionChange(2);
console.log(performance.toString());
```
I get:
```bash
ERROR: subset.proportionChange is not a function
    TypeError: subset.proportionChange is not a function
        at sync (/Users/lgomez/Projects/etl/index.js:104:32)
        at process._tickCallback (internal/process/next_tick.js:68:7)
```
So I checked [this](https://github.com/data-forge/data-forge-ts/blob/a2efde6e7bd4c0246b3ab7f1c9b57491597c9f2e/src/lib/series.ts#L273) and and realized this is a method of Series. So I tried:
```javascript
const syms = `ADA,BCH,BTC,BTG,BTS,DASH,EOS,ETC,ETH`.split(',')
const subset = prices.subset(syms).take(10)
console.log(subset.toString());

const series = subset.getSeries('BTC');
console.log(series.toString());

const performance = series.proportionChange(2);
console.log(performance.toString());
```
And got:
```bash
__index__   ADA   BCH   BTC      BTG   BTS       DASH     EOS   ETC      ETH
----------  ----  ----  -------  ----  --------  -------  ----  -------  -------
2017-06-16  null  null  2481.74  null  0.318329  165.654  null  18.4334  371.358
2017-06-17  null  null  2591.41  null  0.349656  169.012  null  21.3327  373.975
2017-06-18  null  null  2568.92  null  0.335961  177.434  null  22.1432  377.562
2017-06-19  null  null  2578.25  null  0.330724  198.481  null  22.9697  371.813
2017-06-20  null  null  2657.32  null  0.331358  203.353  null  21.6293  376.359
2017-06-21  null  null  2709.33  null  0.32211   185.504  null  20.4588  338.292
2017-06-22  null  null  2714.58  null  0.312718  183.122  null  20.4527  340.326
2017-06-23  null  null  2760.61  null  0.324769  186.924  null  21.3475  340.811
2017-06-24  null  null  2688.44  null  0.327701  185.133  null  20.2919  334.167
2017-06-25  null  null  2627.95  null  0.312531  181.87   null  22.5758  321.283

__index__   __value__
----------  ---------
2017-06-16  2481.74
2017-06-17  2591.41
2017-06-18  2568.92
2017-06-19  2578.25
2017-06-20  2657.32
2017-06-21  2709.33
2017-06-22  2714.58
2017-06-23  2760.61
2017-06-24  2688.44
2017-06-25  2627.95

__index__   __value__
----------  ---------------------
2017-06-17  0.04419076937954825
2017-06-18  -0.008678673000412819
2017-06-19  0.0036318764305622313
2017-06-20  0.030668088819936067
2017-06-21  0.019572351090572367
2017-06-22  0.0019377484470330304
2017-06-23  0.016956582602096898
2017-06-24  -0.026142772792969695
2017-06-25  -0.02250003719629236
```
Which makes more sense... now... as per the guide here: https://github.com/data-forge/data-forge-ts/blob/master/docs/guide.md#rolling-window Rolling window is supposed to work for DataFrame and Series but when I try to run it directly on the `prices` DataFrame, it throws. I then tried looping through `prices` like so:
```javascript
    for (var column in subset.getColumns()) {
      const name = column.name;
      console.log('column: %o', name);
      const series = column.series;
      console.log(series.toString());

      // const series = subset.getSeries('BTC');
      // console.log(series.toString());

      // const performance = series.proportionChange(2);
      // console.log(performance.toString());
    }
```
but get:
```bash
column: undefined
ERROR: Cannot read property 'toString' of undefined
    TypeError: Cannot read property 'toString' of undefined
        at sync (/Users/lgomez/Projects/etl/index.js:108:26)
        at process._tickCallback (internal/process/next_tick.js:68:7)
```
But if I try:
```javascript
    subset.forEach(column => {
      console.log('column: %o', column);
    })
```
I get:
```bash
column: { ADA: null,
  BCH: null,
  BTC: '2481.74',
  BTG: null,
  BTS: '0.318329',
  DASH: '165.654',
  EOS: null,
  ETC: '18.4334',
  ETH: '371.358' }
column: { ADA: null,
  BCH: null,
  BTC: '2591.41',
  BTG: null,
  BTS: '0.349656',
  DASH: '169.012',
  EOS: null,
  ETC: '21.3327',
  ETH: '373.975' }
....
```
So the data IS there...

**In other words... If you wanted to run a rollingWindow calculation on a DataFrame (not a single Series), how would you do it?**

Here's the CSV of this same sample data used here if you'd like to try it. I'm running out of ideas.

```csv
ADA,BCH,BTC,BTG,BTS,DASH,EOS,ETC,ETH
,,2481.74,,0.318329,165.654,,18.4334,371.358
,,2591.41,,0.349656,169.012,,21.3327,373.975
,,2568.92,,0.335961,177.434,,22.1432,377.562
,,2578.25,,0.330724,198.481,,22.9697,371.813
,,2657.32,,0.331358,203.353,,21.6293,376.359
,,2709.33,,0.32211,185.504,,20.4588,338.292
,,2714.58,,0.312718,183.122,,20.4527,340.326
,,2760.61,,0.324769,186.924,,21.3475,340.811
,,2688.44,,0.327701,185.133,,20.2919,334.167
,,2627.95,,0.312531,181.87,,22.5758,321.283
```

Happy to contribute to the docs once I get this working.

Thank you!

**Comment by ashleydavis** [comment]
Actually sorry proportionChange and percentChange only work on series.

So you need to extract the series from the dataframe before you can do that:

`const performance = df.getSeries("SomeSeries").proportionChange();`

Or

`const performance = df.deflate(row => row.SomeSeries).proportionChange();`

Let me know if that works for you.

**Comment by ashleydavis** [comment]
Here are the relevant API docs:

https://data-forge.github.io/data-forge-ts/classes/series.html#proportionchange

https://data-forge.github.io/data-forge-ts/classes/series.html#percentchange

**Comment by lgomez** [comment]
Any idea on why looping through the DataFrame may not be working?

**Comment by ashleydavis** [comment]
Please feel free to contribute to any of the code or docs. I accept pull requests.

**Comment by lgomez** [comment]
I mean, any obvious errors?

**Comment by ashleydavis** [comment]
Can you please share a gist with me including code and some sample data? I'll take a look at it and figure out why it isn't working.

**Comment by lgomez** [comment]
Thank you... Here you go: https://gist.github.com/lgomez/6634df6eb37f5acd2f6bb3dac3aa0317.

Sorry if I stop responding but it's 2am here and I'm passing out.

Thank you for your help!

**Comment by ashleydavis** [comment]
Haha, get some sleep. It's night time here as well.

I'll try and have a look at it tomorrow.

**Comment by ashleydavis** [comment]
Ok this had me perplexed but now I see the problem in your code.

You are using the wrong kind of for loop for this kind of occasion.

You are doing this:

```
for (var column in prices.getColumns()) {
  var name = column.name;
  console.log('name: %o', name);
  var series = column.series;
  console.log('series: %o', series);
}
```

You should be doing this:

```
for (var column of prices.getColumns()) {
  var name = column.name;
  console.log('name: %o', name);
  var series = column.series;
  console.log('series: %o', series);
}
```
Note replacing `in` keyword with `of`.

Here's an explanation of the differences:

https://www.typescriptlang.org/docs/handbook/iterators-and-generators.html

**Comment by ashleydavis** [comment]
Also you can simplify the start of your code:

```
const prices = data
  .withIndex(row => moment(row.date).format('YYYY-MM-DD'))
  .dropSeries("date");

```
To this:

```
const prices = data
  .parseDates("date", "YYYY-MM-DD")
  .dropSeries("date");

```

**Comment by lgomez** [comment]
Oh wow! 20 years of JS work and I never before came across the difference between `for...in` and `for...of`. This made my day. Yes it works. Thank you.

And as for `.parseDates("date", "YYYY-MM-DD")`... Thank you. One dependency less. Always a good thing. I'll keep data wrangling ;) and let you know how it goes.

**Comment by lgomez** [comment]
Ashley,

I've been playing around with data today and have one question:

What's your recommended way for transforming every value of every column in a DataFrame?

I thought `.select` on a DataFrame would give me that but I could not get that to work so I am now doing this:
```javascript
    const transform = (df,fn) => {
      let transformed = df
      for (const column of df.getColumns()) {
        const { name, series } = column
        const values = series.select(value => fn(value))
        transformed = transformed.withSeries(name, values)
      }
      return transformed
    }
```
Which I call like this:
```javascript
    const zeroFilled = transform(prices, value => {
      if (value === null) return 0
      return value
    })
```
But I have a feeling I just don't know the "right" way of doing it.

**Comment by ashleydavis** [comment]
There may not be a right way of doing it ;)

`transformSeries` or `generateSeries` might be what you are looking for.

I've updated the API docs for both functions today to include some examples.

https://data-forge.github.io/data-forge-ts/classes/dataframe.html#generateseries
https://data-forge.github.io/data-forge-ts/classes/dataframe.html#transformseries

Let me know if either of those functions help.

**Comment by ashleydavis** [comment]
You could do something like this:

function fillZero (value) {
    if (value === undefined || value === null) {
        return 0;
    }

    return value;
}

const columnTransformer = df.getColumns().toObject(column => column.name, column => fillZero);
const modifiedDf = df.transformSeries(columnTransformer);

Haven't tested this code! Let me know if it helps!
